### PR TITLE
BUG: fixes weekday for dates before 1752

### LINKED ISSUE #53738: BUG: weekday incorrect for dates before 0000-03-01
### Pandas version checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.

- [X] I have confirmed this bug exists on the [main branch](https://pandas.pydata.org/docs/dev/getting_started/install.html#installing-the-development-version-of-pandas) of pandas.

### Reproducible Example

```python
In [1]: pd.Timestamp('0000-03-01').weekday()
Out[1]: 2

In [2]: pd.Timestamp('0000-02-29').weekday()
Out[2]: 2
```

### Issue Description

pandas thinks that before Wednesday there was another Wednesday - my guess is that this was written by a French pupil trying to get an extra half-day off? 😄 

### Expected Behavior

I'd expect
```python
In [1]: pd.Timestamp('0000-03-01').weekday()
Out[1]: 2

In [2]: pd.Timestamp('0000-02-29').weekday()
Out[2]: 1
```

At least, that's what Rust's chrono gives:
```Rust
use chrono::{NaiveDate};

fn main() {
    let date_str = "0000-02-29";
    let date = NaiveDate::parse_from_str(date_str, "%Y-%m-%d").unwrap();
    let weekday_abbrev = date.format("%a").to_string();
    println!("date: {:?}", date);
    println!("Weekday (abbreviated): {}", weekday_abbrev);
}
```
outputs
```Rust
date: 0000-02-29
Weekday (abbreviated): Tue
```

Likewise, https://calculat.io/en/date/what-day-of-the-week-is/29-february-0000 gives it as Tuesday

### Installed Versions

<details>

INSTALLED VERSIONS
------------------
commit           : 965ceca9fd796940050d6fc817707bba1c4f9bff
python           : 3.10.6.final.0
python-bits      : 64
OS               : Linux
OS-release       : 5.10.102.1-microsoft-standard-WSL2
Version          : #1 SMP Wed Mar 2 00:30:59 UTC 2022
machine          : x86_64
processor        : x86_64
byteorder        : little
LC_ALL           : None
LANG             : en_GB.UTF-8
LOCALE           : en_GB.UTF-8

pandas           : 2.0.2
numpy            : 1.24.3
pytz             : 2023.3
dateutil         : 2.8.2
setuptools       : 67.6.1
pip              : 23.0.1
Cython           : None
pytest           : 7.3.1
hypothesis       : None
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : None
html5lib         : None
pymysql          : None
psycopg2         : None
jinja2           : 3.1.2
IPython          : 8.13.2
pandas_datareader: None
bs4              : 4.12.2
bottleneck       : None
brotli           : None
fastparquet      : None
fsspec           : None
gcsfs            : None
matplotlib       : 3.7.1
numba            : None
numexpr          : None
odfpy            : None
openpyxl         : None
pandas_gbq       : None
pyarrow          : 12.0.0
pyreadstat       : None
pyxlsb           : None
s3fs             : None
scipy            : None
snappy           : None
sqlalchemy       : None
tables           : None
tabulate         : None
xarray           : None
xlrd             : None
zstandard        : None
tzdata           : 2023.3
qtpy             : None
pyqt5            : None
None

</details>

### PR DESCRIPTION
**Pull request by mcgeestocks**
- [x] closes #53738 (Replace xxxx with the GitHub issue number)
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [x] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

### CODE REVIEW

**Comment by mcgeestocks** [comment]
Approach was to switch to implementing [Gauss's algorithm for weekday](http://berndt-schwerdtfeger.de/wp-content/uploads/pdf/cal.pdf).  Ended up being quite interesting.

**Comment by mcgeestocks** [comment]
Thanks @MarcoGorelli! 

Haven't used Hypothesis before by tried to follow the patterns elsewhere in the codebase. 
Also by "standard library" I assumed you were referring to `datetime`. Let me know if I missed the mark any what you were looking for.

**Comment by MarcoGorelli** [comment]
yup, that's right! one way to do it could be to make a dict with
```python
{0: 'Mon', 1: 'Tue', ...}
```
and then compare `.weekday` with `dt.datetime(...).strftime('%a')`

something like that

File: pandas/tests/scalar/timestamp/test_timestamp.py [hunk 1]
```diff
@@ -223,6 +225,31 @@ def test_resolution(self):
...
+        ],
+    )
+    def test_dow_historic(self, date_string, expected_true, expected_false):
+        # GH 53738
+        dt = Timestamp(date_string)
+        dow = dt.weekday()
+        assert dow == expected_true
+        assert not dow == expected_false
+
+    @given(DATETIME_IN_PD_TIMESTAMP_RANGE_NO_TZ)
```
Comment (MarcoGorelli) [inline pandas/tests/scalar/timestamp/test_timestamp.py:246]: I think we should just use `st.datetimes` (where `st` is hypothesis.strategies), and then:
- convert that to a pandas Timestamp
- compare Timestamp.weekday with the standard library's .weekday()

`DATETIME_IN_PD_TIMESTAMP_RANGE_NO_TZ` only goes from about 1700 to about 2300 - I think there's a separate issue about that. But anyway, until that's fixed, let's use the full range of what the standard library offers

**Review by MarcoGorelli — CHANGES_REQUESTED** [review]
nice one! just got a comment on the parametric test

File: pandas/tests/scalar/timestamp/test_timestamp.py (hunk 1 above)
Comment (mcgeestocks) [reply pandas/tests/scalar/timestamp/test_timestamp.py]: Done!

File: doc/source/whatsnew/v2.1.0.rst [hunk 2]
```diff
@@ -364,9 +364,11 @@ Datetimelike
 - Bug in :func:`date_range` when ``freq`` was a :class:`DateOffset` with ``nanoseconds`` (:issue:`46877`)
 - Bug in :meth:`Timestamp.date`, :meth:`Timestamp.isocalendar`, :meth:`Timestamp.timetuple`, and :meth:`Timestamp.toordinal` were returning incorrect results for inputs outside those supported by the Python standard library's datetime module (:issue:`53668`)
 - Bug in :meth:`Timestamp.round` with values close to the implementation bounds returning incorrect results instead of raising ``OutOfBoundsDatetime`` (:issue:`51494`)
+- Bug in :meth:`Timestamp.weekday` with days incorrectly calculating before year 1752 (:issue:`53738`)
```
Comment (MarcoGorelli) [inline doc/source/whatsnew/v2.1.0.rst:367]: "Bug in :meth:`Timestamp.weekday` was returning incorrect results before year 1752 (:issue:`53738`)"

File: pandas/tests/scalar/timestamp/test_timestamp.py [hunk 3]
```diff
@@ -223,6 +227,31 @@ def test_resolution(self):
...
+            ("1752-01-01", 5, 6),
+            ("2023-06-18", 6, 0),
+        ],
+    )
+    def test_dow_historic(self, date_string, expected_true, expected_false):
+        # GH 53738
+        dt = Timestamp(date_string)
+        dow = dt.weekday()
+        assert dow == expected_true
+        assert not dow == expected_false
```
Comment (MarcoGorelli) [inline pandas/tests/scalar/timestamp/test_timestamp.py:246]: why have both?  just `expected` and `assert result == expected` should be enough

File: pandas/tests/scalar/timestamp/test_timestamp.py [hunk 4]
```diff
@@ -223,6 +227,31 @@ def test_resolution(self):
         assert dt.as_unit("ms").resolution == Timedelta(milliseconds=1)
         assert dt.as_unit("s").resolution == Timedelta(seconds=1)
 
+    @pytest.mark.parametrize(
+        "date_string, expected_true, expected_false",
+        [
+            ("0000-2-29", 1, 2),
```
Comment (MarcoGorelli) [inline pandas/tests/scalar/timestamp/test_timestamp.py:233]: can we include a negative date too?

File: pandas/tests/scalar/timestamp/test_timestamp.py [hunk 5]
```diff
@@ -223,6 +227,31 @@ def test_resolution(self):
...
+        dt = Timestamp(date_string)
+        dow = dt.weekday()
+        assert dow == expected_true
+        assert not dow == expected_false
+
+    @given(st.datetimes())
+    def test_dow_sanity(self, dt):
+        # GH 53738
+        result = Timestamp(dt).weekday()
+        expected = datetime(dt.year, dt.month, dt.day).weekday()
```
Comment (MarcoGorelli) [inline pandas/tests/scalar/timestamp/test_timestamp.py:252]: can we just have `expected = dt.weekday()`?

File: pandas/tests/scalar/timestamp/test_timestamp.py [hunk 6]
```diff
@@ -223,6 +227,31 @@ def test_resolution(self):
...
+            ("0000-3-1", 2, 3),
+            ("1582-10-14", 3, 4),
+            ("1582-10-15", 4, 5),
+            ("1752-01-01", 5, 6),
+            ("2023-06-18", 6, 0),
+        ],
+    )
+    def test_dow_historic(self, date_string, expected_true, expected_false):
+        # GH 53738
+        dt = Timestamp(date_string)
```
Comment (MarcoGorelli) [inline pandas/tests/scalar/timestamp/test_timestamp.py:243]: `dt` is often used as an alias for the stdlib `datetime` (`import datetime as dt`) - can we use `ts` instead (for "timestamp")? same in the test below

File: pandas/tests/scalar/timestamp/test_timestamp.py [hunk 7]
```diff
@@ -223,6 +227,31 @@ def test_resolution(self):
...
+        dow = dt.weekday()
+        assert dow == expected_true
+        assert not dow == expected_false
+
+    @given(st.datetimes())
+    def test_dow_sanity(self, dt):
```
Comment (MarcoGorelli) [inline pandas/tests/scalar/timestamp/test_timestamp.py:249]: `parametric` instead of `sanity`?

**Review by MarcoGorelli — CHANGES_REQUESTED** [review]
generally looks good, just got quite a few minor comments

File: pandas/tests/scalar/timestamp/test_timestamp.py (hunk 4 above)
Comment (mcgeestocks) [reply pandas/tests/scalar/timestamp/test_timestamp.py:233]: Ah for BCE right. Sorry I miss understood what you meant by negative before.  

My current implementation of Gauss's isn't happy with negative dates, will work on fixing that.

Comment (mcgeestocks) [reply pandas/tests/scalar/timestamp/test_timestamp.py]: I'm having a hard time finding a valid source for expected values to confirm that my implementation of Gauss's algorithm is performing as expected for negative dates.

I tried the built-in `datetime` lib to compare against' but it didn't' seem to like negative date-strings. 

Is the a source of truth that you'd recommend?

Comment (MarcoGorelli) [reply pandas/tests/scalar/timestamp/test_timestamp.py]: I'd suggest polars:
```python
import polars as pl
pl.Series(['-2000-01-01']).str.to_date()
```
It's not an optional dependency (though perhaps we should consider it? that's for a separate issue anyway) so we can't use it in the parametric tests, but we can use it to generate expected values by hand for this unit test

Comment (MarcoGorelli) [reply pandas/tests/scalar/timestamp/test_timestamp.py]: looks like you've done it correctly anyway 🙌

**Review by MarcoGorelli — APPROVED** [review]
Looks good to me, thanks @mcgeestocks !

I've just moved this to the 2.0.3 milestone, as technically it's a regression introduced by version 2.0.0 (before, `.dt.weekday` was always correct, as timestamps were only supported between about 1700 and 2300)

EDIT: requesting changes for now, see below

File: pandas/tests/scalar/timestamp/test_timestamp.py (hunk 4 above)
Comment (mcgeestocks) [reply pandas/tests/scalar/timestamp/test_timestamp.py]: Thanks @MarcoGorelli appreciate the tip re: Polars!

**Comment by MarcoGorelli** [comment]
CI failure is unrelated, but probably safer to wait til that's addressed before merging

**Review by MarcoGorelli — CHANGES_REQUESTED** [review]
Although this looks correct, the implementation looks a bit different to what's linked to in https://en.wikipedia.org/wiki/Determination_of_the_day_of_the_week#Gauss's_algorithm

could you explain what the various letters you've introduced (e.g. `e`) refer to please?

EDIT: seems like you've used the implementation from http://berndt-schwerdtfeger.de/wp-content/uploads/pdf/cal.pdf, which looks good. All good then, I'll add a note linking to that

File: pandas/_libs/tslibs/ccalendar.pyx [hunk 8]
```diff
@@ -103,16 +113,25 @@ cdef int dayofweek(int y, int m, int d) noexcept nogil:
...
+        long c
+        int g
+        int f
+        int e
+
+    if (m < 3):
+        y -= 1
+
+    c = quot(y, 100)
+    g = y - quot(y, 100) * 100
```
Comment (MarcoGorelli) [inline pandas/_libs/tslibs/ccalendar.pyx:128]: could just use `g = y - 100*c` here, right?

**Review by MarcoGorelli — APPROVED** [review]
Right, I've noted the source of the implementation, have simplified slightly, and have expanded the parametric test to include negative dates (by comparing against numpy)

Very confident about this now - will merge today unless there's any objections, would be good to get this fix into 2.0.3

**Pull request merged**